kotlin_engine/src/main/java/com/github/paylike/kotlin_engine/viewmodel/PaylikeEngine.kt

```diff
@@ -0,0 +1,287 @@
+import com.github.paylike.kotlin_client.domain.dto.tokenize.request.TokenizeTypes
+import com.github.paylike.kotlin_engine.helper.PaylikeEngineError
+import com.github.paylike.kotlin_engine.helper.exceptions.*
+import com.github.paylike.kotlin_engine.model.repository.EngineRepository
+import com.github.paylike.kotlin_engine.model.service.ApiMode
+import com.github.paylike.kotlin_luhn.PaylikeLuhn
+import com.github.paylike.kotlin_money.PaymentAmount
+import com.github.paylike.kotlin_request.exceptions.PaylikeException
+import java.util.*
+import java.util.function.Consumer
+import kotlin.reflect.full.superclasses
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+
+/** Paylike engine */
+// TODO KDoc of engine
+class PaylikeEngine(private val merchantId: String, private val apiMode: ApiMode) : Observable() {
+
+    val repository: EngineRepository = EngineRepository()
+
+    var currentState: EngineState = EngineState.WAITING_FOR_INPUT
+        private set
+
+    var error: PaylikeEngineError? = null
+        private set
+
+    private val apiService: PaylikeClient = PaylikeClient()
+
+    private val log: Consumer<Any> = Consumer { Log.e("Engine logger", it.toString()) }
+
+    /**
+     * Execute api calls and create the necessary data for the [EngineRepository.paymentRepository]
+     */
+    suspend fun createPaymentDataDto(cardNumber: String, cvc: String, month: Int, year: Int) {
+        try {
+            checkValidState(
+                validState = EngineState.WAITING_FOR_INPUT,
+                callerFun = object {}.javaClass.enclosingMethod?.name!!
+            )
+            if (apiMode == ApiMode.LIVE && !PaylikeLuhn.isValid(cardNumber)) {
+                throw InvalidCardNumberException(
+                    "Card number is not valid according to Luhn algorithm"
+                )
+            }
+            val paylikeCardDto: PaylikeCardDto
+            coroutineScope {
+                val cardNumberToken = async {
+                    apiService.tokenize(TokenizeData(TokenizeTypes.PCN, cardNumber))
+                }
+                val cvcToken = async { apiService.tokenize(TokenizeData(TokenizeTypes.PCSC, cvc)) }
+                paylikeCardDto =
+                    PaylikeCardDto(
+                        cardNumberToken.await(),
+                        cvcToken.await(),
+                        ExpiryDto(month, year)
+                    )
+            }
+            repository.paymentRepository =
+                PaymentData(
+                    card = paylikeCardDto,
+                    integration = PaymentIntegrationDto(this.merchantId),
+                )
+        } catch (e: Exception) {
+            setErrorState(e)
+            this.notifyObservers()
+        }
+    }
+
+    /** Start function for a payment flow */
+    suspend fun startPayment(paymentAmount: PaymentAmount, paymentTestDto: PaymentTestDto?) {
```

Comment:
`startPayment` should consume the final payment body dto, with maybe the paymentTestDto (It can be argued that the test DTO should be passed here as param, or already included in the body payment object)

However, what we do currently absolutely stops the user to create subscriptions for example. In a subscription it is not guaranteed that you gonna have a root level `PaymentAmount`. You could simply have it in the `plan` field.

Reply:
I refactored this functionality to be completely modular! now we can set the datafields independently from each other (except for the two compulsory integration and card data)